operator-common/src/main/java/io/strimzi/operator/common/model/PemKeyStoreSupplier.java

```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.operator.common.model;
+
+/**
+ * Interface to be implemented for returning a PEM KeyStore
+ */
+public interface PemKeyStoreSupplier {
```

Comment:
You're right that it doesn't provide a keystore directly. The thing that I do like about the current name is that it makes it clear what it is intended to be used for. Although `PemPeerIndentity` is a more accurate description, I wonder if it wouldn't be as clear to people trying to use it. What about something more like `PemClientAuthIdentity`? Then `PemTrustSet` for the other one like you suggested elsewhere.

Taking into account your other comment about the keystore methods that the KafkaAgentClient uses, perhaps we could update the "Supplier" class to also have a jksKeyStore() method. If we did that we could drop the `Pem` from the name.

In terms of interface vs record vs class, although currently `ClusterOperatorKeyStoreSupplier` and `UserOperatorKeyStoreSupplier` have different logic to parse the `Secret`, looking at them again I think the logic could be common, it's just the keys in the Secret data that differs.  So I wonder if having `ClientAuthIdentity` as an abstract class makes sense? Then the concrete classes would be `ClusterOperatorClientAuthIdentity` and `UserOperatorClientAuthIdentity` (both quite long names) which each defines the secret key to use for the lookup.

I'll take a look at addressing the null/empty comments. I didn't follow what you meant about the `Optional`, are you suggesting the create method for the suppliers (or whatever we call them) should return an optional? I wonder if the constructor should throw an error if the parsing fails, although I'm not sure what ripple effect that will have on the tests, so will take a look.

Reply:
Is the `Client` part really, needed in `PemClientAuthIdentity`? Since a bundle of cert, key and CA cert chain can also be used to identify a server when doing mTLS I suspect we have a `PemAuthIdentity` the root of your abstract class hierarchy, with `PemClientAuthIdentity` as a subtype, and then the CO and UO classes as subtypes of that, if that would be useful. I'm not sure whether we really need distinct `ClusterOperatorClientAuthIdentity` and `UserOperatorClientAuthIdentity` types though. The difference between them appears to me mainly to be how they're constructed (i.e. which Secrets and what the data is expected to contain), rather than wanting to constrain what you can pass instances to, or a requirement for method overloading. This is why I suggested using factory methods. But if you really want separate classes I'm fine with that.

In any case I guess I'm happy with those names you suggest.

My point about Optional was you that with the interface you have to rely on documenting that the key shouldn't be null if the cert is not null, and vice versa. This is a possibility you can completely exclude with a class (because you can enforce rules in the constructor), which simplifies life for callers.